# Storing datset in Python
In this project, we'll begin learning how to clean data using Python. We'll perform data cleaning on a real-world data set of artworks contained in the Museum of Modern Art (MoMA).

Here are a few takeaways you can expect by the end of this lesson:

1.How to work with and prepare text data in Python
2.Basic data cleaning techniques
3.How to correct errors during data cleaning

We'll learn more about the MoMA dataset we'll be working with on the next screen, but for now let's take a moment to understand how Python stores datasets.

# Reading the Dataset

Each row of the dataset represents a unique piece of art. Let's take a look at the first five rows:

In [1]:
import csv

with open("artworks.csv", encoding="utf-8") as file:
    reader = csv.reader(file)
    rows = list(reader)
    header=rows[:1]
    print(header)


[['Title', 'Artist', 'Nationality', 'BeginDate', 'EndDate', 'Gender', 'Date', 'Department']]


A list represents each row of our data, with each item in that list representing an item from the row. Each "row list" is stored in one single outer list — which is why we call this format a list of lists. The diagram above shows a simple table of data and its list of lists representation.

# Determing the length of the dataset

In [2]:
moma=rows[1:]
len(moma)

16725

# Cleaning the Nationality and Gender Columns

Notice how the Nationality and Gender columns have parentheses (()) at the beginning and the end of the values. Our task in this screen is to remove those values.
Our task, however, is to remove a substring, not replace it. In order to remove a substring, we use a simple trick: we replace the substring with an empty string: "".

In [3]:
print(moma[1][2])
print(moma[1][5])

(Spanish)
(Male)


In [4]:
for row in moma:
    nationality=row[2]
    nationality=nationality.replace("(","")
    nationality=nationality.replace(")","")
    row[2]=nationality
    #For Gender:
    gender=row[5]
    gender=gender.replace("(","")
    gender=gender.replace(")","")
    row[5]=gender

In [5]:
print(moma[1][2])
print(moma[1][5])

Spanish
Male


# String Capitalization

In [6]:
def view(table,row_no):
    count = 0
    for row in table:
        if "male" in row[row_no] or "" == row[row_no]:
            count += 1
    return count
view(moma,5)

3235

So, there are 3235 items in the gender column which has issues null values and capitalization.

In [7]:
#For Gender
for row in moma:
    gender=row[5]
    gender=gender.title()
    if not gender:
        gender="Gender Unknown/Other"
    row[5]=gender

In [8]:
#For Nationality
for row in moma:
    nationality=row[2]
    nationality=nationality.title()
    if not nationality:
        nationality="Nationality Unknown"
    row[2]=nationality

In [9]:
view(moma,2)

0

So far, we have converted all the data in lowercase to uppercase and filled in "unknown" where null values were present.

# Errors During Data Cleaning

Having analyzed the artist nationalities, we'll now look at the BeginDate and EndDate columns. These columns contain the birth and death dates of the artist who created the work. Let's look at the contents of the first five rows in each column:

In [10]:
for row in moma[:5]:
    birth_date = row[3]
    death_date = row[4]
    print([birth_date, death_date])

['(1947)', '(2013)']
['(1916)', '(2007)']
['(1870)', '(1943)']
['(1861)', '(1944)']
['(1857)', '(1927)']


In [11]:
def clean_and_convert(date):
    if date!="":
        date=date.replace("(","")
        date=date.replace(")","")
        date=int(date)
    return date
for row in moma:
    birth_date=row[3]
    death_date=row[4]
    birth_date=clean_and_convert(birth_date)
    death_date=clean_and_convert(death_date)
    row[3]=birth_date
    row[4]=death_date

In [12]:
print(moma[:1])

[['Dress MacLeod from Tartan Sets', 'Sarah Charlesworth', 'American', 1947, 2013, 'Female', '1986', 'Prints & Illustrated Books']]


# Parsing Numbers from Complex Strings, Part One

On the previous screen, we converted the BeginDate and EndDate columns into numeric values. If we were to combine the data from the BeginDate column — which represents the artist's year of birth — with the data in the Date column — which represents the year when the artist created the piece of art — we can calculate the age at which the artist produced the work.

our task will be to remove all the extra characters so that we're left with either a range or a single year. Then, we'll finish processing the data on the screens that follow. For the two special cases listed above, we'll do the following:

Where there is an 's that signifies a decade, we'll use the year without the 's.
Where c. or similar indicates an approximate year, we'll remove the c. but keep the year.

In [13]:
bad_chars = ["(", ")", "c", "C", ".", "s", "S"]

def strip_characters(string):
    string = str(string)
    for char in bad_chars:
        string = string.replace(char, "")
    return string


def clean_dataset(dataset):
    for row in dataset:
        date = row[6]
        cleaned_date = strip_characters(date)
        row[6] = cleaned_date
        if cleaned_date.isnumeric():
            row[6]=int(cleaned_date)
# def clean_dataset(dataset):
#     bad_chars = ["(", ")", "c", "C", ".", "s", "S"]
#     for row in dataset:
#         # Clean Date column
#         date = row[6]
#         cleaned_date = strip_characters(date)
#         row[6] = cleaned_date
#         # Cast Date column to int
#         if cleaned_date.isnumeric():
#             row[6] = int(cleaned_date)
for row in moma[:20]:
    print(row[6])
clean_dataset(moma)

1986
1978
1889-1911
1927-1940
1903
1957
1924
1978-1983
2001
(1941)
1949-1950
1963
1908-1911
(1934)
1997
(1931-1933)
1972
1967
1923-1924
1979


# Parsing Numbers from Complex Strings, Part Two

In [14]:
def process_date(date):
    date = str(date).strip("'")
    if "-" in date:
        split_date = date.split("-")
        date_one = split_date[0]
        date_two = split_date[1]
        date = (int(date_one) + int(date_two)) / 2
        date = round(date)
    else:
        date = int(date)
    return date


for row in moma:
    date = row[6]
    cleaned_date = process_date(date)
    row[6] = int(cleaned_date)

for row in moma[:20]:
    print(row[6])

clean_dataset(moma)



1986
1978
1900
1934
1903
1957
1924
1980
2001
1941
1950
1963
1910
1934
1997
1932
1972
1967
1924
1979


In [15]:
print(moma[:6])

[['Dress MacLeod from Tartan Sets', 'Sarah Charlesworth', 'American', 1947, 2013, 'Female', 1986, 'Prints & Illustrated Books'], ['Duplicate of plate from folio 11 verso (supplementary suite, plate 4) from ARDICIA', 'Pablo Palazuelo', 'Spanish', 1916, 2007, 'Male', 1978, 'Prints & Illustrated Books'], ['Tailpiece (page 55) from SAGESSE', 'Maurice Denis', 'French', 1870, 1943, 'Male', 1900, 'Prints & Illustrated Books'], ['Headpiece (page 129) from LIVRET DE FOLASTRIES, À JANOT PARISIEN', 'Aristide Maillol', 'French', 1861, 1944, 'Male', 1934, 'Prints & Illustrated Books'], ['97 rue du Bac', 'Eugène Atget', 'French', 1857, 1927, 'Male', 1903, 'Photography'], ['Pictorial ornament (folio 11) from WOODCUTS', 'Antonio Frasconi', 'American', 1919, 2013, 'Male', 1957, 'Prints & Illustrated Books']]


# Creating a new clean csv file

In [23]:
import csv

# Open a new file to write the cleaned dataset
with open("cleaned_dataset", "w", newline="", encoding="utf-8") as outfile:

    # Create a CSV writer object
    writer = csv.writer(outfile)

    # Write the header row to the CSV file
    for d in header:
        writer.writerow(d)

    # Write each row of the cleaned dataset to the CSV file
    for row in moma:
        writer.writerow(row)
